components/places/src/bookmark_sync/store.rs

```diff
@@ -443,6 +445,21 @@ impl<'a> BookmarksStore<'a> {
     }
 
     fn apply_remote_items(&self, now: Timestamp) -> Result<()> {
+        // Remove all keywords from old and new URLs, and remove new keywords
+        // from all existing URLs. The `NOT NULL` conditions are important; they
+        // ensure that SQLite uses our partial indexes, instead of a table scan.
```

Comment:
Yeah! We could maybe enable query plan logging by default in tests? There's also SQLite's [`INDEXED BY`](https://www.sqlite.org/lang_indexedby.html), which fails the query if it can't use the index you specify. The docs strongly discourage using it, but testing feels like the perfect place to do that. We could also add some test-only hooks that use `sql_support::query_plan::QueryPlan` to generate one, then verify that it mentions the indices.

Reply:
Filed #2781.